llvm/lib/SYCLLowerIR/MutatePrintfAddrspace.cpp

```diff
@@ -0,0 +1,251 @@
+                false, false)
+
+// Public interface to the SYCLMutatePrintfAddrspacePass.
+ModulePass *llvm::createSYCLMutatePrintfAddrspaceLegacyPass() {
+  return new SYCLMutatePrintfAddrspaceLegacyPass();
+}
+
+PreservedAnalyses
+SYCLMutatePrintfAddrspacePass::run(Module &M, ModuleAnalysisManager &MAM) {
+  Type *Int8Type = Type::getInt8Ty(M.getContext());
+  auto *CASLiteralType = PointerType::get(Int8Type, ConstantAddrspaceID);
+  Function *CASPrintfFunc = getCASPrintfFunction(M, CASLiteralType);
+
+  FunctionVecTy FunctionsToDrop;
+  bool ModuleChanged = false;
+  for (Function &F : M) {
+    if (!F.isDeclaration())
+      continue;
+    if (!F.getName().startswith("_Z18__spirv_ocl_printf"))
+      continue;
+    if (F.getArg(0)->getType() == CASLiteralType)
+      // No need to replace the literal type and its printf users
+      continue;
+    ModuleChanged |=
+        setFuncCallsOntoCASPrintf(&F, CASPrintfFunc, FunctionsToDrop);
+  }
+  for (Function *F : FunctionsToDrop)
+    F->eraseFromParent();
+
+  return ModuleChanged ? PreservedAnalyses::all() : PreservedAnalyses::none();
+}
+
+/// Helper implementations
+namespace {
+
+/// Get the constant addrspace version of the __spirv_ocl_printf declaration,
+/// or generate it if the IR module doesn't have it yet. Also make it
+/// variadic so that it could replace all non-variadic generic AS versions.
+Function *getCASPrintfFunction(Module &M, PointerType *CASLiteralType) {
+  Type *Int32Type = Type::getInt32Ty(M.getContext());
+  auto *CASPrintfFuncTy = FunctionType::get(Int32Type, CASLiteralType,
+                                            /*isVarArg=*/true);
+  // extern int __spirv_ocl_printf(
+  //                const __attribute__((opencl_constant)) char *Format, ...)
+  FunctionCallee CASPrintfFuncCallee =
+      M.getOrInsertFunction("_Z18__spirv_ocl_printfPU3AS2Kcz", CASPrintfFuncTy);
+  auto *CASPrintfFunc = cast<Function>(CASPrintfFuncCallee.getCallee());
+  CASPrintfFunc->setCallingConv(CallingConv::SPIR_FUNC);
+  CASPrintfFunc->setDSOLocal(true);
+  return CASPrintfFunc;
+}
+
+/// Generate the constant addrspace version of the generic addrspace-residing
+/// global string. If one exists already, get it from the module.
+Constant *getCASLiteral(GlobalVariable *GenericASLiteral) {
+  Module *M = GenericASLiteral->getParent();
+  // Appending the stable suffix ensures that only one CAS copy is made for each
+  // string. In case of the matching name, llvm::Module APIs will ensure that
+  // the existing global is returned.
+  std::string CASLiteralName = GenericASLiteral->getName().str() + "._AS2";
+  IRBuilder<> Builder(M->getContext());
+  if (GlobalVariable *ExistingGlobal =
+          M->getGlobalVariable(CASLiteralName, /*AllowInternal=*/true))
+    return ExistingGlobal;
+
+  StringRef LiteralValue;
+  getConstantStringInfo(GenericASLiteral, LiteralValue);
+  GlobalVariable *Res = Builder.CreateGlobalString(LiteralValue, CASLiteralName,
```

Comment:
```suggestion
  if (GlobalVariable *ExistingGlobal =
          M->getGlobalVariable(CASLiteralName, /*AllowInternal=*/true))
    return ExistingGlobal;

  StringRef LiteralValue;
  getConstantStringInfo(GenericASLiteral, LiteralValue);
  IRBuilder<> Builder(M->getContext());
  GlobalVariable *Res = Builder.CreateGlobalString(LiteralValue, CASLiteralName,
```

Suggestion is to move `Builder` definition closer to its first use